### PR TITLE
Add a new parameter includeAutomatedEntry to API getContributionStatics

### PR DESCRIPTION
**Pull request by daisy-ycguo**
It's a related patch to https://github.com/zanata/zanata-server/pull/1029.

Add a new parameter to getContributionStatistic API.
The parameter is a boolean type to control whether to include automatic entries of translation into statistic.

### CODE REVIEW

**Comment by seanf** [comment]
@daisy-ycguo :+1: Looks good; we just need to plan the change in the interface.

@aeng @carlosmunoz This interface change will break the server (unless building https://github.com/zanata/zanata-server/pull/1029), but it will also break any clients which call this interface. Do we think there are any such clients?

**Comment by carlosmunoz** [comment]
@seanf Do you mean java clients? or REST clients. In REST, the way I see it it is just an extra query parameter, so clients which call this REST interface will continue to work same as before.

**Comment by seanf** [comment]
@carlosmunoz Java clients. Do we have some reason for thinking this endpoint is not in use yet? It is public, after all. Was it released?

**Comment by seanf** [comment]
@daisy-ycguo This api change seems to break tests in zanata-client:

```
Running org.zanata.rest.client.StatisticsResourceClientTest
Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.133 sec <<< FAILURE! - in org.zanata.rest.client.StatisticsResourceClientTest
testGetContributorStatistics(org.zanata.rest.client.StatisticsResourceClientTest)  Time elapsed: 0.041 sec  <<< ERROR!
com.sun.jersey.api.client.UniformInterfaceException: GET http://192.168.0.216:45035/stats/project/about-fedora/version/master/contributor/pahuang/2014-10-01..2014-11-10 returned a response status of 500 Internal Server Error
        at com.sun.jersey.api.client.WebResource.handle(WebResource.java:686)
        at com.sun.jersey.api.client.WebResource.get(WebResource.java:191)
        at org.zanata.rest.client.StatisticsResourceClient.getContributionStatistics(StatisticsResourceClient.java:108)
        at org.zanata.rest.client.StatisticsResourceClientTest.testGetContributorStatistics(StatisticsResourceClientTest.java:75)
```

**Comment by carlosmunoz** [comment]
Ok, but then wouldn't this be a new version of the API?
I mean, clients using the current version of this library would continue to work, and if they want to make use of this parameter they can increase the version and they would have to change their code.

**Comment by seanf** [comment]
@carlosmunoz it's a breaking change, even if it is easy to fix, so by rights it would mean a new major version. (Our strategy for evolving the API really needs work...)

**Comment by seanf** [comment]
@carlosmunoz Another option would be to create a new interface, leaving the old one alone.  eg StatisticsResource2.  See https://docs.oracle.com/javase/tutorial/java/IandI/nogrow.html

Basically, changing APIs is always a bit ugly in Java.

**Comment by carlosmunoz** [comment]
I'm less concerned about the Java API than I am of the REST one. After all, we have made these kind of breaking changes before.

I do agree we need to start thinking about our API versioning. Do we want to do it for this PR though?

Alex and I will help Daisy fix the failing test in the meanwhile.

File: zanata-common-api/src/main/java/org/zanata/rest/service/StatisticsResource.java [hunk 1]
```diff
@@ -151,7 +154,9 @@ ContributionStatistics getContributionStatistics(
                     @PathParam("projectSlug") String projectSlug,
                     @PathParam("versionSlug") String versionSlug,
                     @PathParam("username") String username,
-                    @PathParam("dateRange") String dateRange
+                    @PathParam("dateRange") String dateRange,
+                    @QueryParam("includeAutomatedEntry")
+                    @DefaultValue("false") Boolean includeAutomatedEntry
```
Comment (seanf) [inline zanata-common-api/src/main/java/org/zanata/rest/service/StatisticsResource.java]: @daisy-ycguo This should be `boolean`, not `Boolean`.

**Comment by seanf** [comment]
We have, but we generally try to find out if anyone is likely to be using
the affected API yet. Do we have any idea of the answer to that for this
endpoint?

On 30 October 2015 at 10:52, Carlos Munoz <notifications@github.com> wrote:

> I'm less concerned about the Java API than I am of the REST one. After
> all, we have made these kind of breaking changes before.
>
> I do agree we need to start thinking about our API versioning. Do we want
> to do it for this PR though?
>
> Alex and I will help Daisy fix the failing test in the meanwhile.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-api/pull/36#issuecomment-152369822>.
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

**Comment by carlosmunoz** [comment]
@seanf Not really. But I'll go out on a limb and say there shouldn't be many (if at all). We don't have a guide for coding java clients using our interfaces or anything like that. And even there, it will only affect those users who choose to upgrade to the latest API, which might reduce the probability even more.

**Comment by seanf** [comment]
@aeng I think you created this method (getContributionStatistics). Do you think anyone but us is using it yet?

**Comment by aeng** [comment]
The best bet would be the original requester: fedora users and openstack
users.


---------------------------------------------

Alex Eng
Globalisation Tools Engineering
DID: +61 3514 8262 <callto:+61+3514+8262>
Mobile: +614 2335 3457 <callto:+614+2335+3457>

Red Hat, Asia-Pacific Pty Ltd
Level 1, 193 North Quay
Brisbane 4000
Office: +61 7 3514 8100 <callto:+61+7+3514+8100>
Fax: +61 7 3514 8199 <callto:+61+7+3514+8199>
Website: www.redhat.com

**Comment by aeng** [comment]
Merged into integration/master. https://github.com/zanata/zanata-api/commit/92b4aecd9d750d115b800db3e362870b9dd45a7d

**Comment by definite** [comment]
Please back port to release (3.8)

**Pull request closed without merging**